Move up xpkg build command to crossplane

lsviben opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

Migrates the [up](https://github.com/upbound/up) `xpkg build` command to Crossplane as part of the devEx improvements.

Some notes about the PR for the reviewers:
Most of the PR is just a copy of the https://github.com/upbound/up/internal/pkg/xpkg to `internal/xpkg/v2` (with some unused files removed, non `build` related), and the copy is in the first commit. The rest of the commits are adjusting the code to Crossplane linting, adding functions to the `build`, and trying to make it more Crossplane then Upbound (mostly in comments).

I skipped for now the `push` command as it has a lot of Upbound specific code, like authenticating and creating repositories, so I felt that that deserved a separate, more readable PR, where some of this code would be made less Upbound-specific. 

Other commands like `dep` `init` will be added separately.

Also I skipped the alpha `batch` and `xpextract` commands as they are alpha. We could add those later on. 

#### How this interacts with the existing commands and code in Crossplane

As we already have a `build`/`push` commands in crossplane(and a couple more), and there is very similar code in `internal/xpkg` initially I tried to merge the code from `up` to that (while keeping the commands separate). And in most cases its ok as it seems that both share the same roots. But it got complicated fast because `up` had a lot of changes on the original code, and the Crossplane code had also its changes, so it was a nightmare to merge. 

Thats why the `up` `xpkg` code was moved into `v2`, as it is anyway planned in the future to replace the current Crossplane commands, due to having a richer feature set. So there will be duplicated code in v2, but its much simpler that way, and we can remove v1 at one point.

#### Which command is added

`build`:
Improved version of the current Crossplane build which allows for packing the manifest togather with the controller/function image and examples.

#### Further improvements

There is a bunch of stuff that can still be done here but I wanted to keep this already big PR simpler:
- deprecating the current `build` command 
- deduplication of code between `v1` and `v2` `internal/xpkg` (partially done)
- adding a pretty logger like the one `up` [uses](https://github.com/upbound/up/tree/main/internal/upterm)
- adding other commands in `up xpkg`

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

You MUST either [x] check or ~strikethrough~ every item in the checklist below.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

Fixes #4657 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- ~[ ] Added `backport release-x.y` labels to auto-backport this PR, if necessary.~
- ~[] Opened a PR updating the [docs], if necessary.~
Docs will be covered in https://github.com/crossplane/crossplane/issues/4671

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### How was this tested

Just ran the command for various packages:
```bash
Commands:
  build configuration
    Build a Configuration package.

  build provider
    Build a Provider package.

  build function
    Build a Function package.

  install configuration <package> [<name>]
    Install a Configuration package.

  install provider <package> [<name>]
    Install a Provider package.

  update configuration <name> <tag>
    Update a Configuration package.

  update provider <name> <tag>
    Update a Provider package.

  push configuration <tag>
    Push a Configuration package.

  push provider <tag>
    Push a Provider package.

  xpkg build
    Build a package, by default from the current directory.

Run "kubectl crossplane <command> --help" for more information on a command.
```

```bash
~/go/test/function-dummy (main) $ kubectl crossplane xpkg build -f package/ --controller=provider-dummy:0.1.0
xpkg saved to /home/lovro/go/test/function-dummy/package/function-dummy-303029fb9515.xpkg
```

lsviben commented:
```
codecov/project Failing after 1s — 65.51% (-4.26%) compared to 89b5fe5
```

Looks like the code coverage is failing as the copied over code is not that well tested. I try to add some tests there

tnthornton commented on cmd/crank/xpkg/dep.go:
```diff
@@ -0,0 +1,217 @@
+// AfterApply constructs and binds context to any subcommands
+// that have Run() methods that receive it.
+func (c *depCmd) AfterApply(kongCtx *kong.Context, p pterm.TextPrinter) error {
+	kongCtx.Bind(pterm.DefaultBulletList.WithWriter(kongCtx.Stdout))
```
@lsviben did you want to inherit pterm for output? We originally added it up for the upbound installation and it seems to have propagated to a few other areas (like this subcommand). I mostly ask since it's unclear if the new crossplane cli wants to use pterm output or if it wants to have the option to take a different path.

tnthornton replied:
Also it should be pretty easy to remove pterm from these subcommands if you want to decrease the scope of what is being pulled over 👍.

lsviben replied:
When I saw pterm there and checked it out I liked how you could setup the style. Just added the skeleton of it so we dont copy Upbound style. But maybe its a good idea to leave it out and add it properly in a separate PR.

tnthornton replied:
>But maybe its a good idea to leave it out and add it properly in a separate PR

Sounds great to me. I think if we leave it in here, it's less obvious that it's being added to the repo and why.

tnthornton commented on internal/xpkg/v2/parser/linter/linter.go:
```diff
@@ -0,0 +1,125 @@
+	errOrFmt = "object did not pass any of the linters with following errors: %s"
+)
+
+// TODO(@tnthornton) update API in crossplane-runtime to use Package interface
+// instead of concreate type and remove this package.
```
Something that might be easier to clean up in a future PR 👍.

FWIW, I believe most of this file is a copy of https://github.com/crossplane/crossplane-runtime/blob/c05e74a4b5cf7ce0c42dedf46f5126e71692958f/pkg/parser/linter.go with the exception of the Package declaration below.

lsviben replied:
I [added](https://github.com/crossplane/crossplane-runtime/pull/553) it already to crossplane-runtime, but slightly different name not to break the crossplane-runtime api. Will deduplicate it later on!

tnthornton commented on internal/xpkg/v2/testdata/config_package.yaml:
```diff
@@ -0,0 +1,313 @@
+apiVersion: apiextensions.crossplane.io/v1
```
I think this test file is from a very long time ago and isn't actually used in tests anymore. Over in up, if I remove it and run `make test` the tests still pass.

tnthornton replied:
Same with the `config_meta.yaml` file above.

tnthornton reviewed (commented):
I'm really excited to see this effort getting kicked off. Given my context for these changes over in upbound/up, I mostly targeted the following:
* Did we bring over more than we needed to (for example dep related functionality and that ties into xpls)?
* Do we want to make some of the same assumptions here that were made in upbound/up?

My comments below are all related to that. Let me know if you have any questions or if there is other context I can provide 👍

lsviben commented:
> I'm really excited to see this effort getting kicked off. Given my context for these changes over in upbound/up, I mostly targeted the following:
> 
> * Did we bring over more than we needed to (for example dep related functionality and that ties into xpls)?
> * Do we want to make some of the same assumptions here that were made in upbound/up?
> 
> My comments below are all related to that. Let me know if you have any questions or if there is other context I can provide 👍

Thanks for taking the time to review @tnthornton ! 

I initially started with adding all the commands, but then removed some because they are alpha (`batch` and `xpextract`), then `push` as I think we need to make it a bit less Upbound-centric. So I was left with these 3.

Looking at it now again, and after the discussion on the SIG meeting, where we confirmed that the most important are `build` and `push` I feel like its a good idea to drop `init` and `dep` from this PR to make it simpler. Especially `dep` which has a lot of code, and is only usefull when we add `xpls`. So ill do that

lsviben commented:
Ok, made the following changes:
- Removed `dep`, `init` and related code from
- Removed `pterm`
- Removed alpha gate
- Bumpred crossplane runtime to include the [Lintable](https://github.com/crossplane/crossplane-runtime/pull/553) interface, which we need for deduplicating the linter stuff. And deduplicated it.

phisco commented on internal/xpkg/v2/build.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// New returns a new Builder.
+func New(pkg, ab, ex parser.Backend, pp parser.Parser, ep *examples.Parser) *Builder {
+	return &Builder{
+		pb: pkg,
+		ab: ab,
+		eb: ex,
+		pp: pp,
+		ep: ep,
+	}
+}
```
some more documentation on the meaning of these would be nice, but we can leave that to a follow up PR 😅

lsviben replied:
Made them a bit more self-explanatory

phisco commented on cmd/crank/xpkg/build.go:
```diff
@@ -0,0 +1,202 @@
+	root    string
+	fetch   fetchFn
+
+	Output       string   `optional:"" short:"o" xor:"xpkg-build-out" help:"Path for package output."`
```
```suggestion
	Output       string   `optional:"" short:"o" help:"Path for package output."`
```

phisco commented:
Checked the output of this against the one from `up` and it's exactly the same as expected:
https://explore.ggcr.dev/?image=docker.io/phisco/test:v0.0.0-provider-nop-with-up
https://explore.ggcr.dev/?image=docker.io/phisco/test:v0.0.0-provider-nop-with-crank

We can definitely refactor it further, but as an starting point and initial porting of the `up` functionalities, I'd say it's ok 👍

haarchri reviewed (approved):
LGTM - thanks 

test successful done! 
`./crank xpkg build --ignore="init/*.yaml,kind-config.yaml" --output=test.xpkg`

haarchri commented on internal/xpkg/v2/testdata/examples/provider.yaml:
```diff
@@ -0,0 +1,21 @@
+metadata:
+  name: provider-aws
+spec:
+  package: crossplane/provider-jet-aws:main
```
nit: remove the jet name

haarchri commented on internal/xpkg/v2/testdata/examples/ec2/instance.yaml:
```diff
@@ -0,0 +1,29 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: ec2.aws.jet.crossplane.io/v1alpha2
```
nit: use exisiting apiVersion for future

haarchri commented on internal/xpkg/v2/testdata/examples/ec2/internetgateway.yaml:
```diff
@@ -0,0 +1,27 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: ec2.aws.jet.crossplane.io/v1alpha2
```
nit: use exisiting apiVersion for future

haarchri commented on internal/xpkg/v2/testdata/examples/ecr/repository.yaml:
```diff
@@ -0,0 +1,26 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: ecr.aws.jet.crossplane.io/v1alpha2
```
nit: use exisiting apiVersion for future

haarchri commented on internal/xpkg/v2/build.go:
```diff
@@ -0,0 +1,321 @@
+	errParseAuth         = "an auth extension was supplied but could not be parsed"
+	errAuthNotAnnotated  = "an auth extension was supplied but but the " + providerConfigKind + " object could not be found"
+	errNotExactlyOneMeta = "package must contain exactly one meta object"
+	authMetaAnno         = "auth.upbound.io/group"
```
is this something we need here ?

jastang replied:
given this is used to configure `ProviderConfig` functionality in `console.upbound.io`, probably not. It does imply that any provider in the future that wants to take advantage of that functionality would have to be built with [up](https://github.com/upbound/up), which is the case today.

lsviben replied:
Removed the auth extension option then. We can leave this to be `up` specific
